### ISSUE TITLE
CreateGPULinearAlgebra: ErrorInvalidSource

### ISSUE THREAD

**Issue by chaseaucoin** [jdermody/brightwire#1]
When CreateGPULinearAlgebra(false) is called as it tries to load the module kernel.cubin I'm getting the following error: "ErrorInvalidSource: This indicates that the device kernel source is invalid." 

Windows 10
Card: GXF 980TI
Cuda toolkit 7.5 is installed. 

Any help would be appreciated.

Thank you

**Comment by jdermody** [comment]
It looks like the included cubin file might not be compatible with Maxwell.

Can you try using [this ptx kernel file](https://github.com/jdermody/brightwire/files/547093/kernel.zip)?  You will need to copy it into the /LinearAlgebra/cuda directory (and choose copy to output directory).

You will also need to change ln. 165 of /LinearAlgebra/CudaProvider from ".cubin" to ".ptx"

**Comment by chaseaucoin** [comment]
That worked perfect. Thank you.

**Comment by chaseaucoin** [comment]
There seems to be a large hit to performance utilizing the GPU on the IRIS sample. Thoughts on cause?

**CPU bound:**

Training a 4x8x3 neural network...
Epoch: 50; t-error: 0.0455 [-0.0006]; time: 0.00s; score: 70.00 %
Epoch: 100; t-error: 0.0297 [-0.0001]; time: 0.00s; score: 70.00 %
Epoch: 150; t-error: 0.0220 [-0.0002]; time: 0.00s; score: 73.33 %
Epoch: 200; t-error: 0.0156 [-0.0001]; time: 0.00s; score: 86.67 %
Epoch: 250; t-error: 0.0116 [-0.0001]; time: 0.00s; score: 96.67 %
Epoch: 300; t-error: 0.0092 [0.0000]; time: 0.00s; score: 96.67 %

Training a 4x8x16x32x16x8x3 neural network...
Epoch: 50; t-error: 0.0688 [0.0000]; time: 0.01s; score: 33.33 %
Epoch: 100; t-error: 0.0686 [0.0000]; time: 0.01s; score: 33.33 %
Epoch: 150; t-error: 0.0369 [-0.0006]; time: 0.01s; score: 70.00 %
Epoch: 200; t-error: 0.0157 [-0.0005]; time: 0.01s; score: 96.67 %
Epoch: 250; t-error: 0.0083 [0.0026]; time: 0.01s; score: 100.00 %
Epoch: 300; t-error: 0.0039 [0.0000]; time: 0.01s; score: 100.00 %

**GPU Bound:**

Training a 4x8x3 neural network...
Epoch: 50; t-error: 0.0455 [-0.0006]; time: 0.05s; score: 70.00 %
Epoch: 100; t-error: 0.0297 [-0.0001]; time: 0.05s; score: 70.00 %
Epoch: 150; t-error: 0.0220 [-0.0002]; time: 0.08s; score: 73.33 %
Epoch: 200; t-error: 0.0156 [-0.0001]; time: 0.05s; score: 86.67 %
Epoch: 250; t-error: 0.0116 [-0.0001]; time: 0.08s; score: 96.67 %
Epoch: 300; t-error: 0.0092 [0.0000]; time: 0.05s; score: 96.67 %

Training a 4x8x16x32x16x8x3 neural network...
Epoch: 50; t-error: 0.0688 [0.0000]; time: 0.14s; score: 33.33 %
Epoch: 100; t-error: 0.0686 [0.0000]; time: 0.18s; score: 33.33 %
Epoch: 150; t-error: 0.0369 [-0.0006]; time: 0.17s; score: 70.00 %
Epoch: 200; t-error: 0.0157 [-0.0005]; time: 0.18s; score: 96.67 %
Epoch: 250; t-error: 0.0083 [0.0026]; time: 0.18s; score: 100.00 %
Epoch: 300; t-error: 0.0038 [0.0000]; time: 0.13s; score: 100.00 %

**Comment by chaseaucoin** [comment]
increasing the batch size seems to have helped.

**Comment by chaseaucoin** [comment]
Looks like that just has to do with the sample set data size. The MNIST sample set doesn't suffer from this. The GPU made a marked improvement in testing times.

**Comment by jdermody** [comment]
Glad it's working now.  I've updated CudaProvider to use the .ptx kernel by default moving forward.

Moving data in and out of the GPU is the overhead. When the cost of the computation itself is minimal the overhead outweighs the benefits.

The batch size is an interesting one - too big and accuracy goes down but too small and the network loses some ability to regularise itself. Larger batch sizes also reduce the training time as the internal matrix multiplications get bigger (and easier to parallelise).
